pkg/util/winutil/pdhutil/pdhcounter.go

```diff
@@ -24,123 +29,222 @@ var (
+	GetAllValues() (map[string]float64, error)
 }
 
-// Initialize initializes a counter set object
-func (p *PdhCounterSet) Initialize(className string, counterName string) error {
+// pdhCounter contains info that is common to all counter types
+type pdhCounter struct {
+	handle PDH_HCOUNTER
 
-	// refresh PDH object cache (refresh will only occur periodically)
-	_, _ = tryRefreshPdhObjectCache()
+	// Parts of PDH counter path
+	ObjectName   string // also referred to by Microsoft as a counterset, class, or performance object
+	InstanceName string
+	CounterName  string
 
-	p.className = className
-	p.counterName = counterName
+	initError error
+}
 
-	pdherror := pfnPdhOpenQuery(uintptr(0), uintptr(0), &p.query)
-	if ERROR_SUCCESS != pdherror {
-		err := fmt.Errorf("Failed to open PDH query handle %#x", pdherror)
-		return err
-	}
-	return nil
+// pdhEnglishCounter implements AddToQuery for both single and multi instance english counters
+type pdhEnglishCounter struct {
+	pdhCounter
 }
 
-// GetEnglishCounterInstance returns a specific instance of the given counter
-// the className and counterName must be in English.
-// See PdhAddEnglishCounter docs for details
-// https://learn.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhaddenglishcountera
-func GetEnglishCounterInstance(className string, counterName string, instance string) (*PdhSingleInstanceCounterSet, error) {
+// pdhEnglishSingleInstanceCounter is a specialization for single-instance counters
+// https://learn.microsoft.com/en-us/windows/win32/perfctrs/about-performance-counters
+type PdhEnglishSingleInstanceCounter struct {
+	pdhEnglishCounter
+}
 
-	var p PdhSingleInstanceCounterSet
-	if err := p.Initialize(className, counterName); err != nil {
-		return nil, err
+// pdhMultiInstanceCounterSet is a specialization for a multi-instance counters
+// https://learn.microsoft.com/en-us/windows/win32/perfctrs/about-performance-counters
+type PdhEnglishMultiInstanceCounter struct {
+	pdhEnglishCounter
+	verifyfn CounterInstanceVerify
+}
+
+func (counter *pdhCounter) ShouldInit() bool {
+	if counter.handle != PDH_HCOUNTER(0) {
+		// already initialized
+		return false
 	}
+	return true
+}
 
-	path, err := pfnPdhMakeCounterPath("", className, instance, counterName)
+// Implements PdhCounter.AddToQuery for english counters.
+func (counter *pdhEnglishCounter) AddToQuery(query *PdhQuery) error {
+	path, err := pfnPdhMakeCounterPath("", counter.ObjectName, counter.InstanceName, counter.CounterName)
 	if err != nil {
-		return nil, fmt.Errorf("Failed to make counter path %s: %v", counterName, err)
-	}
-	pdherror := pfnPdhAddEnglishCounter(p.query, path, uintptr(0), &p.counter)
-	if ERROR_SUCCESS != pdherror {
-		return nil, fmt.Errorf("Failed to add english counter %#x", pdherror)
+		counter.initError = fmt.Errorf("Failed to make counter path (\\%s(%s)\\%s): %v", counter.ObjectName, counter.InstanceName, counter.CounterName, err)
```

Comment:
Error is not actually accurate in the sense of implied path because it will be different when instance exists or not

Reply:
I thought it would be helpful to log the args. How would you prefer it represented?

Reply:
I think I am just nitpicking here. Theoretically you would use with or without instance component path name, But doing this makes logging unnecessary complex. It is fine.